---
title: Identification of a DNA methylation signature in blood cells from persons with
  Down Syndrome
output:
  pdf_document: default
  html_notebook: default
---
Matteo Bolner , University of Bologna

Reference article: https://www.ncbi.nlm.nih.gov/pubmed/25701644

From the abstract:  
Down Syndrome (DS) is characterized by a wide spectrum of clinical signs, which include segmental premature aging of central nervous and immune systems. Although it is well established that the causative defect of DS is the trisomy of chromosome 21, the molecular bases of its phenotype are still largely unknown. We used the Infinium HumanMethylation450 BeadChip to investigate DNA methylation patterns in whole blood from 29 DS persons, using their relatives (mothers and unaffected siblings) as controls. This family-based model allowed us to monitor possible confounding effects on DNA methylation patterns deriving from genetic and environmental factors. Although differentially methylated regions (DMRs) displayed a genome-wide distribution, they were enriched on chromosome 21.



Project goal:  
The goal of this project is the development of a pipeline to analyze the DNA methylation data obtained with the Infinium HumanMethylation450 BeadChip from eigth people, four of which are affected by Down Syndrome, and understand if there are significant differences in DNA methylation between the two groups.


**STUDENT 2**

STEP 3      ->      address : 10633381  
STEP 5      ->      detPvalue threshold : 0.05  
STEP 7      ->      normalization : preprocessNoob  
STEP 11     ->      test : ANOVA correcting for female 

## STEPS OF THE ANALYSIS PIPELINE

### 1. Load raw data with minfi and create an object called RGset storing the RGChannelSet object

```{r message=FALSE, warning=FALSE, paged.print=FALSE}
rm(list=ls())

library(minfi)
library(knitr)
library(kableExtra)
library(gap)
library(qqman)
library(gplots)
```

```{r}
load("/home/pelmo/R_wd/Illumina450Manifest_clean.RData")
```


```{r}
targets <- read.metharray.sheet("/home/pelmo/drd_project/data/Input_data/")
RGset <- read.metharray.exp(targets=targets)
```

```{r}
kable(targets) %>% kable_styling("bordered")
```

### 2.  Create the dataframes Red and Green to store the red and green fluorescence respectively

```{r}
Red <- data.frame(getRed(RGset))
Green <- data.frame(getGreen(RGset))
```



### 3. Fill the following table: what are the Red and Green fluorescence for the address assigned to you?  
### Optional: check from the manifest file if the address corresponds to a Type I or a Type II probe and, in case of Type I probe, report its color.

```{r}
my_address <- "10633381"
```

```{r}
Illumina450Manifest_clean_pelmo[Illumina450Manifest_clean_pelmo$AddressA_ID==my_address,]$Infinium_Design_Type
```
The output of the above command returns a null factor; this means that my address doesn't correspond to a probe A.

```{r}
Illumina450Manifest_clean_pelmo[Illumina450Manifest_clean_pelmo$AddressB_ID==my_address,]$Infinium_Design_Type
```
Therefore, my address corresponds to a probe B of type I.

```{r}
Illumina450Manifest_clean_pelmo[Illumina450Manifest_clean_pelmo$AddressB_ID==my_address,]$Color_Channel
```
The color of my type I probe is red.  

```{r}
my_address_red <- t(Red[rownames(Red) %in% my_address,])
my_address_green <- t(Green[rownames(Green) %in% my_address,])

my_address_fluorescences <- merge(my_address_red, my_address_green, by=0)
colnames(my_address_fluorescences) <- c("Sample", "Red fluor", "Green fluor")

kable(my_address_fluorescences) %>% kable_styling("bordered", full_width = F)
```

### 4. Create the object MSet.raw

```{r}
MSet.raw <- preprocessRaw(RGset)
```
### 5. Perform the following quality checks and provide a brief comment to each step:<br/><br/>
#### • QCplot  
```{r}
qc <- getQC(MSet.raw)
plotQC(qc)
```
As can be seen in the QC plot, all the samples appear to be of acceptable quality, with no sample below the cutoff line.
<br/><br/>
    
#### • Check the intensity of negative controls using minfi
```{r}
control_probes <- data.frame(getProbeInfo(RGset, type= "Control"))

control_probes_type <- table(control_probes$Type)
control_probes_type <- as.data.frame(control_probes_type)
colnames(control_probes_type) <- c("Control type", "Freq")
control_probes_type <- control_probes_type[order(-control_probes_type["Freq"]),]
rownames(control_probes_type) <- NULL

#kable(control_probes_type) %>% kable_styling("bordered", full_width = F)%>% scroll_box(height = "200px")

controlStripPlot(RGset, controls="NEGATIVE")
```
The log2 intensities of the negative controls are within the detection range for all of the samples
<br/><br/>  

#### • Calculate detection pValues; for each sample, how many probes have a detection p-value higher than the threshold assigned?  


```{r}
pval_threshold <- 0.05
detp_value <- detectionP(RGset)
```


```{r}
failed <- (detp_value > pval_threshold)
failed_df <- t(data.frame(apply(failed,2,table)))
samples_df <- rownames(failed_df)
rownames(failed_df) <- NULL
failed_df <- cbind(samples_df, failed_df)
colnames(failed_df) <- c("Samples", "pValue < 0.05", "pValue > 0.05")
kable(failed_df) %>% kable_styling("bordered", full_width = F) %>% column_spec(column = 3, bold = T)
```

```{r}
failed_overall <- rowMeans(failed)
which_failed <- (failed_overall>0.01)
length(which_failed)
bad_probes <- names(which_failed[which_failed==TRUE])
length(bad_probes)
```
1905 probes have a detection pValue higher than 0.05 in more than 1% of the samples.

<br></br>

### 6. Calculate raw beta and M values and plot the densities of mean methylation values, dividing the samples in DS and WT.


```{r}
beta <- getBeta(MSet.raw)
m <- getM(MSet.raw)
```

```{r}
#kable(targets) %>% kable_styling("bordered")
wt_samples <- targets[targets$Group=="WT",]
wt_samples_names <- sub("/home/pelmo/drd_project/data/Input_data/", "", wt_samples$Basename)
wt_samples_names
ds_samples <- targets[targets$Group=="DS",]
ds_samples_names <- sub("/home/pelmo/drd_project/data/Input_data/", "", ds_samples$Basename)
ds_samples_names

```

```{r}
samples_names <- colnames(RGset)
wt_beta <- beta[,samples_names %in% wt_samples_names]
wt_m <- m[,samples_names %in% wt_samples_names]
ds_beta <- beta[,samples_names %in% ds_samples_names]
ds_m <- m[,samples_names %in% ds_samples_names]
```


```{r}
mean_wt_beta <- apply(wt_beta,1, mean)
mean_ds_beta <- apply(ds_beta,1, mean)
mean_wt_m <- apply(wt_m,1, mean)
mean_ds_m <- apply(ds_m,1, mean)
```

```{r}
plot(density(mean_wt_beta, na.rm=T), main="Density of mean Beta values", col="blue",xlab="Beta", ylab="Density")
lines(density(mean_ds_beta,na.rm=T), col="red")
legend("topright", legend=c("WT", "DS"), col=c("blue", "red"), lty=1:1)

```

```{r}
plot(density(mean_wt_m, na.rm=T), main="Density of mean M values", col="blue",xlab="Beta", ylab="Density")
lines(density(mean_ds_m, na.rm=T), col="red")
legend("topright", legend=c("WT", "DS"), col=c("blue", "red"), lty=1:1)
```




### 7. Compare raw data with normalized data using the function assigned to each student. Produce a plot with 6 panels (2 rows, 3 columns) in which, for both raw and normalized data, you show the density plots of beta mean values according to the chemistry of the probes, the density plot of beta standard deviation values according to the chemistry of the probes and the boxplot of beta values. Provide a short comment regarding the changes you observe.



The aim of normalization is to remove unavoidable technical variation, in particular systematic bias; out of band probes (OOB) represent measurements of type I probes in the wrong color channel, and are very useful in the estimation of non-specific fluorescence in background correction. With OOB normalization, within-array and between-array significantly reduce artifacts and at the same time improve sensitivity.
Since Infinium I and II probes behave differently, they were divided in the analysis and normalization of the data, as can be seen in the plots below.

```{r}
infinium_I <- Illumina450Manifest_clean_pelmo[Illumina450Manifest_clean_pelmo$Infinium_Design_Type=="I",]
infinium_I <- droplevels(infinium_I)
```

```{r}
infinium_II <- Illumina450Manifest_clean_pelmo[Illumina450Manifest_clean_pelmo$Infinium_Design_Type=="II",]
infinium_II <- droplevels(infinium_II)
```


```{r}
beta_I <- beta[rownames(beta) %in% infinium_I$IlmnID,]
beta_II <- beta[rownames(beta) %in% infinium_II$IlmnID,]
mean_beta_I <- apply(beta_I,1, mean)
mean_beta_II <- apply(beta_II,1, mean)
sd_beta_I <- apply(beta_I, 1, sd)
sd_beta_II <- apply(beta_II, 1, sd)
```

```{r}
preprocess_Noob_data <- preprocessNoob(RGset) 
```

```{r}
preprocess_Noob_beta <- getBeta(preprocess_Noob_data)
preprocess_Noob_beta_I <- preprocess_Noob_beta[rownames(preprocess_Noob_beta) %in% infinium_I$IlmnID,]
preprocess_Noob_beta_II <- preprocess_Noob_beta[rownames(preprocess_Noob_beta) %in% infinium_II$IlmnID,]
mean_preprocess_Noob_beta_I <- apply(preprocess_Noob_beta_I, 1, mean)
mean_preprocess_Noob_beta_II <- apply(preprocess_Noob_beta_II, 1, mean)
sd_preprocess_Noob_beta_I <- apply(preprocess_Noob_beta_I, 1, sd)
sd_preprocess_Noob_beta_II <- apply(preprocess_Noob_beta_II, 1, sd)

```

```{r, fig.height = 10, fig.width = 15}
par(mfrow=c(2,3), mai=c(1.5,0.7,0.5,0.5))

plot(density(mean_beta_I, na.rm=T), col="black", main="Mean of raw Beta values", xlab="Mean of Beta")
lines(density(mean_beta_II, na.rm=T), col=259)
legend("topright", legend=c("I", "II"), col=c("black", 259), lty=1:1)

plot(density(sd_beta_I, na.rm=T), col="black", main="Standard Deviation (SD) of raw Beta values", xlab="SD of Beta")
lines(density(sd_beta_II, na.rm=T), col=259)
legend("topright", legend=c("I", "II"), col=c("black", 259), lty=1:1)

boxplot((beta), main= "Raw Beta values", las = 2)

plot(density(mean_preprocess_Noob_beta_I, na.rm=T), col="black", main="Mean of normalized Beta values", xlab="Mean of Beta")
lines(density(mean_preprocess_Noob_beta_II, na.rm=T), col=259)
legend("topright", legend=c("I", "II"), col=c("black", 259), lty=1:1)

plot(density(sd_preprocess_Noob_beta_I, na.rm=T), col="black", main="Standard Deviation of normalized Beta values", xlab="SD of Beta")
lines(density(sd_preprocess_Noob_beta_II, na.rm=T), col=259)
legend("topright", legend=c("I", "II"), col=c("black", 259), lty=1:1)

boxplot(preprocess_Noob_beta, main="Normalized Beta values", las = 2)

```
The normalization has an immediately noticeable effect on the distribution of beta values: 
  
-The mean of raw Beta values shows that type II probes have lower sensitivity than type I when it comes to the detection of extreme methylation values; after normalization, the peaks are more aligned between probe types, with type II peaks closer to the extreme values. 
  
-The same can be said for the standard deviation; type II probes have higher SD between replicates, and the curve for raw beta values reflects this; after normalization, both curves show higher peaks but lower SD, meaning that the density for lower SDs is higher.  
  
-The box plot after normalization shows significantly lower variation between samples, and much more similar beta value distributions.  
  
Overall, the normalization procedure implemented allows to significantly reduce experimental variability both within and between samples.



<br></br>

### 8. Consider the normalized beta and M values and remove the probes previously defined as bad according to the detection pValue.

```{r}
preprocess_Noob_m <- getM(preprocess_Noob_data)
preprocess_Noob_beta_clean <- preprocess_Noob_beta[!rownames(preprocess_Noob_beta) %in% bad_probes,]
preprocess_Noob_m_clean <- preprocess_Noob_m[!rownames(preprocess_Noob_beta) %in% bad_probes,]
```

### 9. Check beta and M matrixes generated in step 8 for homo/heteroschedasticity; comment the plot. Optional: plot the lowess line.

```{r}
preprocess_Noob_beta_clean_mean <- apply(preprocess_Noob_beta_clean, 1, mean)
preprocess_Noob_beta_clean_sd <- apply(preprocess_Noob_beta_clean, 1, sd)
preprocess_Noob_m_clean_mean <- apply(preprocess_Noob_m_clean, 1, mean)
preprocess_Noob_m_clean_sd <- apply(preprocess_Noob_m_clean, 1, sd)
```

```{r, fig.width=15, fig.height=8}
par(mfrow=c(1,2))
smoothScatter(preprocess_Noob_beta_clean_mean, preprocess_Noob_beta_clean_sd, main="Beta values", xlab="Mean(Beta)", ylab="SD(Beta)")
lines(lowess(preprocess_Noob_beta_clean_mean, preprocess_Noob_beta_clean_sd), col="red")
smoothScatter(preprocess_Noob_m_clean_mean, preprocess_Noob_m_clean_sd, main= "M values", xlab="Mean(M)", ylab="SD(M)")
lines(lowess(preprocess_Noob_m_clean_mean, preprocess_Noob_m_clean_sd), col="red")
```
-The plot for beta values shows heteroscedasticity, with mean beta values going from 0 (absence of methylation) to 1 (complete methylation)
  
-The plot for M values shows homoscedasticity, with mean M values not bound to the 0-1 range but going above and below.
  
The red lines represent locally weighted scatterplot smoothing (LOWESS), which uses locally weighted polynomial regression to return a continuous line.


### 10. Perform a PCA on the beta matrix generated in step 8 to check for batch effects. Comment the plot.
```{r}
pca_beta <- prcomp(t(preprocess_Noob_beta_clean), scale=T)
```


```{r}
phenotype <- read.csv("data/Input_data/Samplesheet_report.csv", sep=",")
print(summary(pca_beta))
plot(pca_beta$x[,1], pca_beta$x[,2], cex=2, xlab="PC1", ylab="PC2", xlim=c(-1000,1000), col=c("red","black")[as.numeric(phenotype$Group)])
#legend("bottomright", legend=levels(phenotype$Group), col=c("red","black"), pch=1)
```
The prcomp function was chosen rather than pcicomp, due to the low number of samples analyzed; the first two principal components (PC) were plotted. Due to the relatively low cumulative proportion of variance between the two PCs (0.48) and the small sample size, it can be concluded that the data is not affected by any batch effect.


### 11. Using the beta matrix generated in step 8, identify differentially methylated probes between group A and group B using the functions assigned to each student.

```{r}
anova_function <- function(x) {
  anova_test <- aov(x ~ phenotype$Group+phenotype$Female) #ANOVA correcting for females
  return(summary(anova_test)[[1]][[5]][1])
}
pval_anova <- apply(preprocess_Noob_beta_clean, 1, anova_function)
```

### 12. Apply multiple test correction and set a significant threshold of 0.05. How many probes do you identify as differentially methylated considering nominal pValues? How many after Bonferroni correction? How many after BH correction?

```{r}
final_anova <- data.frame(preprocess_Noob_beta_clean, pval_anova)
final_anova <- final_anova[order(final_anova$pval_anova),]
pvals_sorted <- final_anova[,9]
pval_corrected_bh <- p.adjust(pvals_sorted, "BH")
pval_corrected_bf <- p.adjust(pvals_sorted, "bonferroni")
final_anova_corrected <- data.frame(final_anova, pval_corrected_bh, pval_corrected_bf)
```



```{r}
significant_probes_no_correction <- final_anova_corrected[final_anova_corrected$pval_anova<=0.05,]
dim(significant_probes_no_correction)
significant_probes_bh <- final_anova_corrected[final_anova_corrected$pval_corrected_bh<=0.05,]
dim(significant_probes_bh)
significant_probes_bf <- final_anova_corrected[final_anova_corrected$pval_corrected_bf<=0.05,]
dim(significant_probes_bf)
```

-Considering nominal pValues, 26615 probes appear to be differentially methylated.  
-After Benjamini & Hochberg (BH) correction, 2482 probes appear to be differentially methylated.  
-After Bonferroni (BF) correction, only 93 probes appear to be differentially methylated.  

### 13. Produce a Manhattan plot and a volcano plot of your data.

```{r}
final_anova_corrected <- data.frame(rownames(final_anova_corrected),final_anova_corrected)
```

```{r}
colnames(final_anova_corrected)[1] <- "IlmnID"
final_anova_corrected_annotated <- merge(final_anova_corrected, Illumina450Manifest_clean_pelmo, by="IlmnID")
```



```{r}
manhattan_plot <- data.frame(final_anova_corrected_annotated$CHR, final_anova_corrected_annotated$MAPINFO, final_anova_corrected_annotated$pval_anova)
manhattan_plot$final_anova_corrected_annotated.CHR <- factor(manhattan_plot$final_anova_corrected_annotated.CHR, levels=c((1:22),"X","Y"))
colnames(manhattan_plot) <- c("CHR", "MAPINFO", "PVAL_ANOVA")
chrs <- replace(as.character(manhattan_plot$CHR), as.character(manhattan_plot$CHR)=="X", 23)
chrs <- replace(chrs,chrs=="Y",24)
chrs <- as.numeric(chrs)
manhattan_plot$CHR <- chrs
```

```{r, fig.height = 5, fig.width = 10}

#ALTERNATIVE MANHATTAN PLOT

#palette <- c("red","blue","green","cyan","black","gray","magenta","red","blue","green","cyan","black","gray","magenta","red","blue","green","cyan","black","gray","brown","red","blue","green")
#mhtplot(manhattan_plot, control=mht.control(colors = palette, type = "p", cex = 0.2,cutoffs = -log10(0.05), gap = 10000, labels=chromosomes))
#title("P values of differently methylated loci")
#axis(2,cex=0.5)
#abline(a=-log10(0.05),b=0)

```

```{r, fig.height = 5, fig.width = 13}
library(qqman)
manhattan(manhattan_plot, chr="CHR",bp="MAPINFO",p="PVAL_ANOVA", genomewideline = -log10(0.05), suggestiveline = FALSE, cex=0.5, cex.axis=0.6)
```



```{r}
beta_volcano <- final_anova_corrected[,2:9]
beta_volcano_WT <- beta_volcano[,phenotype$Group=="WT"]
beta_volcano_WT_mean <- apply(beta_volcano_WT, 1, mean)
beta_volcano_DS <- beta_volcano[,phenotype$Group=="DS"]
beta_volcano_DS_mean <- apply(beta_volcano_DS, 1, mean)

delta <- beta_volcano_DS_mean-beta_volcano_WT_mean
```

```{r}
volcano_plot <- data.frame(delta, -log10(final_anova_corrected$pval_anova))
plot(volcano_plot[,1], volcano_plot[,2], pch=16, cex=0.3, xlab= "Mean(Beta(DS))- Mean(Beta(WT))", ylab="-log10(P-value)")
abline(a=-log10(0.05), b=0, col="red")
hypermethylation <- volcano_plot[abs(volcano_plot[,1]>0.2) & volcano_plot[,2]>(-log10(0.05)),]
points(hypermethylation[,1], hypermethylation[,2], pch=16, cex=0.3, col="red")
hypomethylation <- volcano_plot[abs(volcano_plot[,1]<(-0.2)) & volcano_plot[,2]>(-log10(0.05)),]
points(hypomethylation[,1], hypomethylation[,2], pch=16, cex=0.3, col="blue")
legend( x="topright", 
        legend=c("Hypermethylation", "Hypomethylation"), 
        col=c("red","blue"), lwd=1, lty=c(NA,NA), 
        pch=c(20,20), merge=FALSE )

```

### 14. Produce a heatmap of the top 100 differentially methylated probes.


```{r,fig.height = 5, fig.width = 7}
final_anova_corrected_annotated <- final_anova_corrected_annotated[order(final_anova_corrected_annotated$pval_anova),]
heatmap_matrix <- as.matrix(final_anova_corrected_annotated[1:100,2:9])
sample_color <-  vector(mode="character", length=length(phenotype$Group))
for (sample in seq(length(phenotype$Group))) {
    if (phenotype$Group[sample]=="WT") {
        sample_color[sample] <- "green"
    }
    else {
        sample_color[sample] <- "orange"
    }
}
heatmap.2(heatmap_matrix,col=terrain.colors(100),Rowv=T,Colv=T,dendrogram="both",key=T,ColSideColors=sample_color,density.info="none",trace="none",scale="none",symm=F)

```

```{r,fig.height = 5, fig.width = 7}
heatmap.2(heatmap_matrix,col=terrain.colors(100),Rowv=T,Colv=T,hclustfun = function(x) hclust(x,method = 'single'),dendrogram="both",key=T,ColSideColors=sample_color,density.info="none",trace="none",scale="none",symm=F)

```

```{r,fig.height = 5, fig.width = 7}
heatmap.2(heatmap_matrix,col=terrain.colors(100),Rowv=T,Colv=T,hclustfun = function(x) hclust(x,method = 'average'),dendrogram="both",key=T,ColSideColors=sample_color,density.info="none",trace="none",scale="none",symm=F)

```


### Optional
### As DS is caused by the trisomy of chromosome 21, try also to plot the density of the methylation values of the probes mapping on chromosome 21. Do you see a very clear difference between the samples? How many differentially methylated probes do you find on chromosome 21?



```{r}

chr_21_betas <- final_anova_corrected_annotated[,c(1:9,23)]
rownames(chr_21_betas) <- NULL
chr_21_betas <- chr_21_betas[chr_21_betas$CHR==21,]
wt_samples_names_2 <- paste("X", wt_samples_names, sep ="")
ds_samples_names_2 <- paste("X", ds_samples_names, sep= "")
wt_chr_21_betas <- chr_21_betas[,wt_samples_names_2]
ds_chr_21_betas <- chr_21_betas[,ds_samples_names_2]
wt_chr_21_betas_means <- apply(wt_chr_21_betas, 1, mean)
ds_chr_21_betas_means <- apply(ds_chr_21_betas, 1, mean)
plot(density(wt_chr_21_betas_means), main = "Mean Beta values in chromosome 21", col="blue", xlab="Mean Beta value")
lines(density(ds_chr_21_betas_means), col="red")
legend("topright",lty=1:1,col=c("blue", "red"),legend=c("WT","DS"))
```
There appears to be no significant difference in methylation between samples.

```{r}
significant_probes_ids <- rownames(significant_probes_no_correction)
significant_probes_chr_21 <- chr_21_betas$IlmnID %in% significant_probes_ids
length(significant_probes_chr_21)
table(significant_probes_chr_21)
significant_probes_chr_21_ids <- chr_21_betas$IlmnID[significant_probes_chr_21==TRUE]
chr_21_proportion <- (length(significant_probes_chr_21_ids)/length(significant_probes_chr_21))*100

chr_21_proportion 
```
412 probes of the 4237 present on chromosome 21 (9.72%) appear differentially methylated, with a p-value lower than 0.05.
In order to understand if this value is significant, it must be compared with the global proportion.

```{r}
overall_chr_proportions <- (length(significant_probes_ids)/length(Illumina450Manifest_clean_pelmo$IlmnID))*100
overall_chr_proportions
```
The proportion of differentially methylated probes on chromosome 21 is almost double with respect to the overall proportion, indicating that a significant amount of methylation activity on chromosome 21 seems to be correlated with Down Syndrome.














