---
title: "Identification of a DNA methylation signature in blood cells from persons with Down Syndrome"
output: html_notebook
---

## ABSTRACT (COSTRUISCI INTRODUZIONE BREVE)
Down Syndrome (DS) is characterized by a wide spectrum of clinical signs, which include segmental premature aging of central nervous and immune systems. Although it is well established that the causative defect of DS is the trisomy of chromosome 21, the molecular bases of its phenotype are still largely unknown. We used the Infinium HumanMethylation450 BeadChip to investigate DNA methylation patterns in whole blood from 29 DS persons, using their relatives (mothers and unaffected siblings) as controls. This family-based model allowed us to monitor possible confounding effects on DNA methylation patterns deriving from genetic and environmental factors. Although differentially methylated regions (DMRs) displayed a genome-wide distribution, they were enriched on chromosome 21. DMRs mapped in genes involved in developmental functions, including embryonic development (HOXA family) and haematological (RUNX1 and EBF4) and neuronal (NCAM1) development. Moreover, genes involved in the regulation of chromatin structure (PRMD8, KDM2B, TET1) showed altered methylation. The data also showed that several pathways are affected in DS, including PI3K-Akt signaling. In conclusion, we identified an epigenetic signature of DS that sustains a link between developmental defects and disease phenotype, including segmental premature aging.


STUDENT **2**

STEP 3 -> address = 10633381  
STEP 5 -> detPvalue threshold = 0.05  
STEP 7 -> normalization = preprocessNoob  
STEP 11 -> test = Anova correcting for female (vedi rob polverelli)

## STEPS OF THE ANALYSIS PIPELINE

### 1. Load raw data with minfi and create an object called RGset storing the RGChannelSet object (see Script Lesson3_bis_2019)

```{r message=FALSE, warning=FALSE, paged.print=FALSE}
rm(list=ls())

library(minfi)
library(knitr)
library(kableExtra)
```

```{r}
load("/home/pelmo/R_wd/Illumina450Manifest_clean.RData")
```


```{r}
targets <- read.metharray.sheet("/home/pelmo/drd_project/data/")
RGset <- read.metharray.exp(targets=targets)
```

```{r}
kable(targets) %>% kable_styling("bordered")
RGset
```

### 2.  Create the dataframes Red and Green to store the red and green fluorescence respectively

```{r}
Red <- data.frame(getRed(RGset))
Green <- data.frame(getGreen(RGset))
```
```{r}
kable(head(Red)) %>% kable_styling("bordered") %>% scroll_box(width = "100%")
kable(head(Green)) %>% kable_styling("bordered") %>% scroll_box(width = "100%")

dim(Red)
dim(Green)
```



### 3. 3. Fill the following table: what are the Red and Green fluorescence for the address assigned to you? (suggestion: subset the Red and Green dataframes by setting rownames==”the Address you want”). Optional: check from the manifest file if the address corresponds to a Type I or a Type II probe and, in case of Type I probe, report its color.

```{r}
my_address <- "10633381"
```

##qua sotto uso [aaaa,] senza niente dopo virgola perche sto selezionando [rows,columns] e lasciando vuoto scelgo tutte le columns

```{r}
Illumina450Manifest_clean_pelmo[Illumina450Manifest_clean_pelmo$AddressA_ID==my_address,]$Infinium_Design_Type
```
The output of the above command returns a null factor; this means that my address doesn't correspond to a probe A.

```{r}
Illumina450Manifest_clean_pelmo[Illumina450Manifest_clean_pelmo$AddressB_ID==my_address,]$Infinium_Design_Type
```
Therefore, my address corresponds to a probe B of type I.

```{r}
Illumina450Manifest_clean_pelmo[Illumina450Manifest_clean_pelmo$AddressB_ID==my_address,]$Color_Channel

```
The color of my type I probe is red.  

#Just in case I will need it, I obtain the illumina ID corresponding to the CpG locus 
 
```{r}
#my_address_id <- Illumina450Manifest_clean_pelmo[Illumina450Manifest_clean_pelmo$AddressB_ID==my_address,]$IlmnID
#my_address_id
```


```{r}
my_address_red <- t(Red[rownames(Red) %in% my_address,])


my_address_green <- t(Green[rownames(Green) %in% my_address,])

my_address_green
my_address_red

my_address_fluorescences <- merge(my_address_red, my_address_green, by=0)
colnames(my_address_fluorescences) <- c("Sample", "Red fluor", "Green fluor")
kable(my_address_fluorescences) %>% kable_styling("bordered", full_width = F)
```

#a <- getProbeInfo(RGset)
#altro metodo per fare step 3 capire colore e tipo probe

### 4. Create the object MSet.raw

```{r}
MSet.raw <- preprocessRaw(RGset)
```

```{r}
kable(head(getMeth(MSet.raw))) %>% kable_styling("bordered") %>% scroll_box(width="100%")
```

### 5. Perform the following quality checks and provide a brief comment to each step:
• QCplot  

```{r}
qc <- getQC(MSet.raw)
plotQC(qc)
```
As can be seen in the QC plot, all the samples appear to be of acceptable quality, with no sample below the cutoff line.


• Check the intensity of negative controls using minfi  
```{r}
control_probes <- data.frame(getProbeInfo(RGset, type= "Control"))

control_probes_type <- table(control_probes$Type)
control_probes_type <- as.data.frame(control_probes_type)
colnames(control_probes_type) <- c("Control type", "Freq")
control_probes_type <- control_probes_type[order(-control_probes_type["Freq"]),]
rownames(control_probes_type) <- NULL

kable(control_probes_type) %>% kable_styling("bordered", full_width = F)%>% scroll_box(height = "250px")
controlStripPlot(RGset, controls="NEGATIVE")
```
The log2 intensities of the negative controls are within the detection range for all of the samples




• Calculate detection pValues; for each sample, how many probes have a detection p-value higher than the threshold assigned to each student?  


```{r}
pval_threshold <- 0.05
detp_value <- detectionP(RGset)
```


```{r}
failed <- (detp_value > pval_threshold)
failed_df <- t(data.frame(apply(failed,2,table)))
samples_df <- rownames(failed_df)
rownames(failed_df) <- NULL
failed_df <- cbind(samples_df, failed_df)
colnames(failed_df) <- c("Samples", "pValue < 0.05", "pValue > 0.05")
kable(failed_df) %>% kable_styling("bordered", full_width = F)
```

```{r}
failed_overall <- rowMeans(failed)
which_failed <- (failed_overall>0.01)
length(which_failed)
bad_probes <- names(which_failed[which_failed==TRUE])
head(bad_probes)
length(bad_probes)
```
1905 probes out of ???485512??? have a detection pValue higher than 0.05 in more than 1% of the samples.


### 6. Calculate raw beta and M values and plot the densities of mean methylation values, dividing the samples in DS and WT (suggestion: subset the beta and M values matrixes in order to retain DS or WT subjects and apply the function mean to the 2 subsets).


```{r}
beta <- getBeta(MSet.raw)
m <- getM(MSet.raw)
```

```{r}
kable(targets) %>% kable_styling("bordered")
wt_samples <- targets[targets$Group=="WT",]
wt_samples
wt_samples_names <- sub("/home/pelmo/drd_project/data//Input_data/", "", wt_samples$Basename)
wt_samples_names
ds_samples <- targets[targets$Group=="DS",]
#ds_samples
ds_samples_names <- sub("/home/pelmo/drd_project/data//Input_data/", "", ds_samples$Basename)
ds_samples_names

```

```{r}
samples_names <- colnames(RGset)
wt_beta <- beta[,samples_names %in% wt_samples_names]
wt_m <- m[,samples_names %in% wt_samples_names]
ds_beta <- beta[,samples_names %in% ds_samples_names]
ds_m <- m[,samples_names %in% ds_samples_names]

#head(wt_m)
#head(ds_m)
#head(wt_beta)
#head(ds_beta)
```


```{r}
mean_wt_beta <- apply(wt_beta,1, mean)
mean_ds_beta <- apply(ds_beta,1, mean)
mean_wt_m <- apply(wt_m,1, mean)
mean_ds_m <- apply(ds_m,1, mean)
```

```{r}
plot(density(mean_wt_beta, na.rm=T), main="Density of mean Beta values", col="blue",xlab="Beta", ylab="Density")
lines(density(mean_ds_beta,na.rm=T), col="red")
legend("topright", legend=c("WT", "DS"), col=c("blue", "red"), lty=1:1)

```

```{r}
plot(density(mean_wt_m, na.rm=T), main="Density of mean M values", col="blue",xlab="Beta", ylab="Density")
legend("topright", legend=c("WT", "DS"), col=c("blue", "red"), lty=1:1)
lines(density(mean_ds_m, na.rm=T), col="red")


```




### 7. Compare raw data with normalized data using the function assigned to each student (see script Lesson_4_2019). Produce a plot with 6 panels (2 rows, 3 columns) in which, for both raw and normalized data, you show the density plots of beta mean values according to the chemistry of the probes, the density plot of beta standard deviation values according to the chemistry of the probes and the boxplot of beta values. Provide a short comment regarding the changes you observe.



The aim of normalization is to remove unavoidable technical variation, in particular systematic bias; out of band probes (OOB) represent measurements of type I probes in the wrong color channel, and are very useful in the estimation of non-specific fluorescence in background correction. With OOB normalization, within-array and between-array significantly reduce artifacts and at the same time improve sensitivity.
Since Infinium I and II probes behave differently, they were divided in the analysis and normalization of the data, as can be seen in the plots below.

```{r}
infinium_I <- Illumina450Manifest_clean_pelmo[Illumina450Manifest_clean_pelmo$Infinium_Design_Type=="I",]
infinium_I <- droplevels(infinium_I)
```

```{r}
infinium_II <- Illumina450Manifest_clean_pelmo[Illumina450Manifest_clean_pelmo$Infinium_Design_Type=="II",]
infinium_II <- droplevels(infinium_II)
```


```{r}
beta_I <- beta[rownames(beta) %in% infinium_I$IlmnID,]
beta_II <- beta[rownames(beta) %in% infinium_II$IlmnID,]
mean_beta_I <- apply(beta_I,1, mean)
mean_beta_II <- apply(beta_II,1, mean)
sd_beta_I <- apply(beta_I, 1, sd)
sd_beta_II <- apply(beta_II, 1, sd)
```

```{r}
preprocess_Noob_data <- preprocessNoob(RGset)
preprocess_Noob_data
```

```{r}
preprocess_Noob_beta <- getBeta(preprocess_Noob_data)
preprocess_Noob_beta_I <- preprocess_Noob_beta[rownames(preprocess_Noob_beta) %in% infinium_I$IlmnID,]
preprocess_Noob_beta_II <- preprocess_Noob_beta[rownames(preprocess_Noob_beta) %in% infinium_II$IlmnID,]
mean_preprocess_Noob_beta_I <- apply(preprocess_Noob_beta_I, 1, mean)
mean_preprocess_Noob_beta_II <- apply(preprocess_Noob_beta_II, 1, mean)
sd_preprocess_Noob_beta_I <- apply(preprocess_Noob_beta_I, 1, sd)
sd_preprocess_Noob_beta_II <- apply(preprocess_Noob_beta_II, 1, sd)

```

```{r, fig.height = 10, fig.width = 15}
par(mfrow=c(2,3), mai=c(1.5,0.7,0.5,0.5))

plot(density(mean_beta_I, na.rm=T), col="black", main="Mean of raw Beta values", xlab="Mean of Beta")
lines(density(mean_beta_II, na.rm=T), col=259)
legend("topright", legend=c("I", "II"), col=c("black", 259), lty=1:1)

plot(density(sd_beta_I, na.rm=T), col="black", main="Standard Deviation (SD) of raw Beta values", xlab="SD of Beta")
lines(density(sd_beta_II, na.rm=T), col=259)
legend("topright", legend=c("I", "II"), col=c("black", 259), lty=1:1)

boxplot((beta), main= "Raw Beta values", las = 2)

plot(density(mean_preprocess_Noob_beta_I, na.rm=T), col="black", main="Mean of normalized Beta values", xlab="Mean of Beta")
lines(density(mean_preprocess_Noob_beta_II, na.rm=T), col=259)
legend("topright", legend=c("I", "II"), col=c("black", 259), lty=1:1)

plot(density(sd_preprocess_Noob_beta_I, na.rm=T), col="black", main="Standard Deviation of normalized Beta values", xlab="SD of Beta")
lines(density(sd_preprocess_Noob_beta_II, na.rm=T), col=259)
legend("topright", legend=c("I", "II"), col=c("black", 259), lty=1:1)

boxplot(preprocess_Noob_beta, main="Normalized Beta values", las = 2)

```
The normalization has an immediately noticeable effect on the distribution of beta values:  
-The mean of raw Beta values shows that type II probes have lower sensitivity than type I when it comes to the detection of extreme methylation values; after normalization, the peaks are more aligned between probe types, with type II peaks closer to the extreme values.  
-The same can be said for the standard deviation; type II probes have higher SD between replicates, and the curve for raw beta values reflects this; after normalization, both curves show higher peaks but lower SD, meaning that the density for lower SDs is higher.  
-The box plot after normalization shows significantly lower variation between samples, and much more similar beta value distributions.  
Overall, the normalization procedure implemented allows to significantly reduce experimental variability both within and between samples.





### 8. Consider the normalized beta and M values and remove the probes previously defined as bad according to the detection pValue (see script Lesson5_2019)

```{r}
preprocess_Noob_m <- getM(preprocess_Noob_data)
preprocess_Noob_beta_clean <- preprocess_Noob_beta[!rownames(preprocess_Noob_beta) %in% bad_probes,]
preprocess_Noob_m_clean <- preprocess_Noob_m[!rownames(preprocess_Noob_beta) %in% bad_probes,]
length(bad_probes)
identical((dim(preprocess_Noob_m)[1]-dim(preprocess_Noob_m_clean)[1]), length(bad_probes))
identical((dim(preprocess_Noob_beta)[1]-dim(preprocess_Noob_beta_clean)[1]), length(bad_probes))

```

### 9. Check beta and M matrixes generated in step 8 for homo/heteroschedasticity (see script Lesson5_2019); comment the plot. Optional: plot the lowess line.

```{r}
preprocess_Noob_beta_clean_mean <- apply(preprocess_Noob_beta_clean, 1, mean)
preprocess_Noob_beta_clean_sd <- apply(preprocess_Noob_beta_clean, 1, sd)
preprocess_Noob_m_clean_mean <- apply(preprocess_Noob_m_clean, 1, mean)
preprocess_Noob_m_clean_sd <- apply(preprocess_Noob_m_clean, 1, sd)

```

```{r, fig.width=10}
par(mfrow=c(1,2))
smoothScatter(preprocess_Noob_beta_clean_mean, preprocess_Noob_beta_clean_sd, main="Beta values", xlab="Mean(Beta)", ylab="SD(Beta)")
lines(lowess(preprocess_Noob_beta_clean_mean, preprocess_Noob_beta_clean_sd), col="red")
smoothScatter(preprocess_Noob_m_clean_mean, preprocess_Noob_m_clean_sd, main= "M values", xlab="Mean(M)", ylab="SD(M)")
lines(lowess(preprocess_Noob_m_clean_mean, preprocess_Noob_m_clean_sd), col="red")
```
-The plot for beta values shows heteroscedasticity, with mean beta values going from 0 (absence of methylation) to 1 (complete methylation)
-The plot for M values shows homoscedasticity, with mean M values not bound to the 0-1 range but going above and below.
The red lines represent locally weighted scatterplot smoothing (LOWESS), which uses locally weighted polynomial regression to return a continuous line.


### 10. Perform a PCA on the beta matrix generated in step 8 to check for batch effects (see script Lesson6_2019). Comment the plot.
```{r}
pca_beta <- prcomp(t(preprocess_Noob_beta_clean), scale=T)
```


```{r}
phenotype <- read.csv("data/Input_data/Samplesheet_report.csv", sep=",")
print(summary(pca_beta))
str(pca_beta)
pca_beta$x
plot(pca_beta$x[,1], pca_beta$x[,2], cex=2, xlab="PC1", ylab="PC2", xlim=c(-1000,1000), col=c("red","black")[as.numeric(phenotype$Group)])
#text(pca_beta$x[,1], pca_beta$x[,2], labels=rownames(pca_beta$x), cex=0.5, pos=2)
legend("bottomright", legend=levels(phenotype$Group), col=c("red","black"), pch=1)
```
The prcomp function was chosen rather than pcicomp, due to the low number of samples analyzed; the first two principal components (PC) were plotted. Due to the relatively low cumulative proportion of variance between the two PCs (0.48) and the small sample size, it can be concluded that the data is not affected by any batch effect.


### 11. Using the beta matrix generated in step 8, identify differentially methylated probes between group A and group B using the functions assigned to each student (it will take several minutes) (see script Lesson5_bis_2019). Note; it can take several minutes; if you encounter any problem you can run the differential methylated analysis only on a subset of probes (for example those on chromosome 1, 18 and 21)

```{r}
anova_function <- function(x) {
  anova_test <- aov(x ~ phenotype$Group+phenotype$Female) #ANOVA correcting for females
  return(summary(anova_test)[[1]][[5]][1])
}
pval_anova <- apply(preprocess_Noob_beta_clean, 1, anova_function)

```

```{r}
final_anova <- data.frame(preprocess_Noob_beta_clean, pval_anova)
final_anova <- final_anova[order(final_anova$pval_anova),]
pvals_sorted <- final_anova[,9]
pval_corrected_bh <- p.adjust(pvals_sorted, "BH")
pval_corrected_bf <- p.adjust(pvals_sorted, "bonferroni")
final_anova_corrected <- data.frame(final_anova, pval_corrected_bh, pval_corrected_bf)
head(final_anova_corrected)
boxplot(final_anova_corrected, )
```

